feat: update kafka automated deprovision expiration logic to make use of instance types config

miguelsorianod opened the pull request:
## Description

Related to https://issues.redhat.com/browse/MGDSTRM-8483

Kafka lifespans for a given instance size are now specified through the `config/kafka-instance-types-configuration.yaml`.
KFM code was still using the environment variable `KAFKA_LIFE_SPAN` to calculate expiration time of existing kafka instances to mark them for deprovisioning.
This PR updates the expiration logic to use lifespan set in `config/kafka-instance-types-configuration.yaml` instead.

Some important additional remarks:
* The `--enable-deletion-of-expired-kafka` flag that controls whether expiration logic is applied or not has been kept. The alternative would be removing this completely.

## Verification Steps
1.
Create a kafka standard instance
Create 3 developer instances
Then update the created_at time in the DB to a time older than 2 days from now for the standard instance and for 2 of the 3 developer instances
Verify that only the 2 developer instances where the created_at time was updated to older than 2 days are marked as deprovision. The other developer instance and the standard instance should
remain without being marked as deprovisioning

2.
Run fleet manager with the --enable-db-debug flag and check that the SQL statements related to the changes are correct

3. 
Repeat test 1 but now run fleet manager with --enable-deletion-of-expired-kafka=false. No instances should be marked as expired at all

## Checklist (Definition of Done)
<!-- Please strikethrough options not relevant using two tildes ~~Text~~. Do not delete non relevant options -->
- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] Documentation added for the feature
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] Required metrics/dashboards/alerts have been added (or PR created).
- [ ] Required Standard Operating Procedure (SOP) is added.
- [ ] JIRA has been created for changes required on the client side

miguelsorianod commented on internal/kafka/internal/services/kafka.go:
```diff
@@ -567,25 +568,54 @@ func (k *kafkaService) DeprovisionKafkaForUsers(users []string) *errors.ServiceE
 
-	db := dbConn.Update("status", constants2.KafkaRequestStatusDeprovision)
+	var existingKafkaRequests []dbapi.KafkaRequest
+	db := dbConn.Where("instance_type IN (?)", kafkaDeletionInstanceTypes).
```
At the moment the implementation only retrieves `developer` instances because we know that those are the only ones that have expiration.
In theory we shouldn't do this as they could have expiration too. It is kind of an "optimization". Do you think we should just make it fully generic without filtering by the instance type?

miguelsorianod replied:
That makes sense :+1: I will implement it

miguelsorianod replied:
I updated the PR to include the additional proposed logic here

miguelsorianod commented:
I will add some tests (where/if it makes sense) after the round of reviews if the approach implemented looks ok

miguelsorianod commented on internal/kafka/internal/services/kafka.go:
```diff
@@ -567,25 +568,54 @@ func (k *kafkaService) DeprovisionKafkaForUsers(users []string) *errors.ServiceE
 	return nil
 }
 
-func (k *kafkaService) DeprovisionExpiredKafkas(kafkaAgeInHours int) *errors.ServiceError {
-	dbConn := k.connectionFactory.New().
-		Model(&dbapi.KafkaRequest{}).
-		Where("instance_type IN (?)", kafkaDeletionInstanceTypes).
-		Where("created_at <= ?", time.Now().Add(-1*time.Duration(kafkaAgeInHours)*time.Hour)).
-		Where("status NOT IN (?)", kafkaDeletionStatuses)
```
Before it was a single update to the DB with some conditions.
Now it is a query to the DB to get the data, some processing on KFM to calculate the changes and then an update to the DB

miguelsorianod replied:
Yes. Look at the implementation.
We need to check the instance size of each kafka instance and then find it in the configuration file

machi1990 replied:
> Yes. Look at the implementation. We need to check the instance size of each kafka instance and then find the information about that size it in the configuration file to get its lifespan value (if any)

Ah right. Thanks. I see it now. I thought the lifespan was attached to the streaming unit and not streaming unit sizes (the array).

miguelsorianod commented:
@machi1990 regarding the changes in app-interface I think the process should be the following one:

For stage:

It is a little bit tricky. I think we should do the rollout in 2 phases:
1. First in this PR we should include the code changes but we should not include the removal of the `KAFKA_LIFE_SPAN` parameter (although we should not use it internally anywhere anymore)
2. Then once the PR changes have rolled out to stage we should remove the `KAFKA_LIFE_SPAN` template parameter from app-interface for stage
3. Finally we should open another PR where we remove the `KAFKA_LIFE_SPAN` parameter

For prod:
1. Merge this PR
2. In app-interface, remove the `KAFKA_LIFE_SPAN` template parameter and update the commit id reference to a one that includes the changes of this PR

machi1990 commented:
> @machi1990 regarding the changes in app-interface I think the process should be the following one:
> 
> For stage:
> 
> It is a little bit tricky. I think we should do the rollout in 2 phases:
> 
> 1. First in this PR we should include the code changes but we should not include the removal of the `KAFKA_LIFE_SPAN` parameter (although we should not use it internally anywhere anymore)
> 2. Then once the PR changes have rolled out to stage we should remove the `KAFKA_LIFE_SPAN` template parameter from app-interface for stage
> 3. Finally we should open another PR where we remove the `KAFKA_LIFE_SPAN` parameter
> 
> For prod:
> 
> 1. Merge this PR
> 2. In app-interface, remove the `KAFKA_LIFE_SPAN` template parameter and update the commit id reference to a one that includes the changes of this PR. The two changes have to be done at the same time

Yes, this approach make sense.

miguelsorianod commented:
I updated the PR with:
* Addressed https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1042#discussion_r879459166
* Added some unit tests and updated tests that were failing to make them pass

machi1990 reviewed (approved):
lgtm
